### PR TITLE
InfluxDB: InfluxQL: query editor: skip fields in metadata queries

### PR DESCRIPTION
**Pull request by gabor**
in the influxdb influxql query editor, the WHERE section is primarily meant to be used with influxdb_tags. but it is possible to use influxdb_fields there too. it works fine for running the query, but it can cause problems for metadata queries:

if you have an influxdb_field in the WHERE part, and then try to add another segment to WHERE by clicking on the [+], we ask influxdb for possible tag_key values in the dropdown, and, because we have an influxdb_field in it, influxdb returns the "expression too complex for metaquery" error.

to fix this, when doing metadata queries, we only send real influxdb_tags to the database, we skip influxdb_fields. 

to differentiate between influxdb_tags and influxdb_fields, when the query editor starts, we ask influxdb for all tag_keys. with this info, we can then filter the tags+fields array to only have tags when doing a metadata query.

how to test:
- `make devenv sources=influxdb`
- wait 20 seconds
- go to explore, choose the `gdev-influxdb-influxql` data source
- in `select measurement` choose `cpu`
- below, in `field(value)` choose `usage_idle`
- click on the [+] next to the WHERE part and manually enter `usage_idle`
- set the filter to form `usage_idle = 42`
- now click on the `+` again. verify that you get a list of tag keys (`cpu` and `host`)
- choose `cpu`
- click on the `select tag value`
- verify that you get a list of options: `cpu-total`, `cpu0`, `cpu1`, `cpu2`, `cpu3`
- choose `cpu1`
- now open the `cpu` dropdown` in the FROM section
- verify that you get 1 option `cpu` (i mean, verify that you do not get zero options)

### CODE REVIEW

**Review by ifrost — APPROVED** [review]
Splendid! It'd be good to add some tests as well 👍

**Comment by gabor** [comment]
> Splendid! It'd be good to add some tests as well 👍

thanks! tests added.

File: public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx [hunk 1]
```diff
@@ -0,0 +1,136 @@
+import React from 'react';
+import { InfluxQuery } from '../../types';
+import InfluxDatasource from '../../datasource';
+import { render, screen, act } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import { Editor } from './Editor';
+import * as mockedMeta from '../../influxQLMetadataQuery';
+
+jest.mock('../../influxQLMetadataQuery', () => {
+  // const orig = jest.requireActual('../../influxQLMetadataQuery');
```
Comment (ifrost) [inline public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx:10]: Is this line needed?

File: public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      .mockReturnValueOnce(Promise.resolve([])),
+  };
+});
+
+beforeEach(() => {
+  // we need to ts-ignore here, the functions in that module
+  // are mocked, so they do have `mockReset`, but typescript
+  // does not know this
+  // @ts-ignore
+  mockedMeta.getTagKeysForMeasurementAndTags.mockClear();
```
Comment (ifrost) [inline public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx:40]: Instead, you can also cast it with `as jest.Mock`

File: public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        type: 'field',
+        params: ['usage_idle'],
+      },
+    ],
+  ],
+  measurement: 'cpudata',
+};
+
+describe('InfluxDB InfluxQL Visual Editor field-filtering', () => {
+  it('should work', async () => {
```
Comment (ifrost) [inline public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx:91]: Could it be a bit more descriptive? :D

File: public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+    // we click the WHERE/cpu button
+    await act(async () => {
+      userEvent.click(screen.getByRole('button', { name: 'cpu' }));
+    });
+
+    // and verify getTagKeysForMeasurementAndTags was called again,
+    // and in the tags-param we did not receive the `field1` part.
+    expect(mockedMeta.getTagKeysForMeasurementAndTags).toHaveBeenCalledTimes(2);
+    // need to ts-ignore the `.mock` part
```
Comment (ifrost) [inline public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx:110]: Would `as jest.Mock` work?

File: public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    // @ts-ignore
+    expect(mockedMeta.getTagKeysForMeasurementAndTags.mock.calls[1][2]).toStrictEqual(ONLY_TAGS);
+
+    // now we click on the WHERE/host2 button
+    await act(async () => {
+      userEvent.click(screen.getByRole('button', { name: 'host2' }));
+    });
+
+    // very `getTagValues` was called once, and in the tags-param we did not receive `field1`
```
Comment (ifrost) [inline public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx:119]: very -> verify? (same below)

**Review by ifrost — APPROVED** [review]
I think it looks okay, just left a few small comments 👍

File: public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx (hunk 1 above)
Comment (gabor) [reply public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx]: good catch, removed

File: public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx (hunk 3 above)
Comment (gabor) [reply public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx]: sorry, you're right. when i created it i just left in a placeholder text, and then forgot to update it.

File: public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx (hunk 5 above)
Comment (gabor) [reply public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx]: right :)

File: public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx (hunk 2 above)
Comment (gabor) [reply public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx]: sure

File: public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx (hunk 4 above)
Comment (gabor) [reply public/app/plugins/datasource/influxdb/components/VisualInfluxQLEditor/Editor.tags.test.tsx]: sure

**Pull request merged**